### PR TITLE
feat: switch to CorpPass Cloud for test form

### PR DESCRIPTION
**Pull request by mantariksh**
## Problem

CorpPass is cutting over to its new cloud infrastructure, and part of the migration process involves a rehearsal on production. While this rehearsal is happening, we don't want any public users to be able to log in to CorpPass, but we still need to be able to test the new cloud infrastructure with our own test form.

## Solution

Switch to CorpPass Cloud only for a specific test form ID.

The cloud migration involves two changes:
- Changing to a new OOB verification endpoint (`CORPPASS_IDP_ENDPOINT` env var)
- Changing to a new CorpPass public key (`CP_IDP_CERT_PATH` env var)

The approach is to get all forms to use the old env vars, while using the new env vars _only_ for a specific form that we choose.

Why does this work? **Outside of the rehearsal hours**, the _old_ CorpPass infrastructure will be up. Hence the _old_ env vars will work, and all public users will be able to use CorpPass normally. However, the _new_ CorpPass infrastructure will be down, so our test form (which is linked to the new env vars) will not work.

Conversely, **during rehearsal hours**, the old CorpPass infrastructure will be down. The old env vars will direct public users to the old CorpPass login endpoint, so their login will fail. On the other hand, the new CorpPass infrastructure will be up, so our test form will work (assuming the migration succeeds).

This implementation creates new env vars which default to the production Cloud values, as well as the form ID of a test form which I created. The defaults allow us to deploy the code without worrying about environment variables, while still allowing us to adjust the variables if needed.

## Screenshots

When public users attempt to log in to CorpPass during the rehearsal hours, they will see the following Toastr when they are redirected back to the form after logging in:
image

## Tests
- [ ] Create, activate and submit a new CorpPass form. Should be able to log in and submit successfully.
- [ ] Log out of this new CorpPass form and deactivate it. Change the `CP_CLOUD_FORM_ID` env var to the ID of the new CorpPass form. Should still be able to activate the form, but should see "There was an unexpected error with your log in" after attempting to log in using CorpPass.
- [ ] Delete the `CP_CLOUD_CERT_PATH`, `CP_CLOUD_ENDPOINT` and `CP_CLOUD_FORM_ID` env vars from staging. You should now be able to log in and submit the new CorpPass form again.

### CODE REVIEW

File: src/app/modules/spcp/spcp.controller.ts [hunk 1]
```diff
@@ -106,14 +116,15 @@ export const addSpcpSessionInfo: RequestHandler<ParamsDictionary> = async (
...
-  const { authType } = (req as WithForm<typeof req>).form
+  const { authType, _id } = (req as WithForm<typeof req>).form
   if (authType !== AuthType.SP && authType !== AuthType.CP) return next()
 
   const jwtResult = SpcpFactory.extractJwt(req.cookies, authType)
   // No action needed if JWT is missing, just means user is not logged in
   if (jwtResult.isErr()) return next()
 
-  return SpcpFactory.extractJwtPayload(jwtResult.value, authType)
+  const useCpCloud = spcpFeature.props?.cpCloudFormId === String(_id)
```
Comment (karrui) [inline src/app/modules/spcp/spcp.controller.ts:126]: add TODO to remove?

File: src/app/modules/spcp/spcp.controller.ts [hunk 2]
```diff
@@ -142,11 +153,14 @@ export const isSpcpAuthenticated: RequestHandler<ParamsDictionary> = (
   res,
   next,
 ) => {
-  const { authType } = (req as WithForm<typeof req>).form
+  const { authType, _id } = (req as WithForm<typeof req>).form
   if (authType !== AuthType.SP && authType !== AuthType.CP) return next()
 
+  const useCpCloud = spcpFeature.props?.cpCloudFormId === String(_id)
```
Comment (karrui) [inline src/app/modules/spcp/spcp.controller.ts:159]: add TODO to remove?

File: src/app/modules/spcp/spcp.controller.ts [hunk 3]
```diff
@@ -221,16 +235,18 @@ export const handleLogin: (
     res.cookie('isLoginError', true)
     return res.redirect(destination)
   }
+  const useCpCloud = spcpFeature.props?.cpCloudFormId === String(form._id)
```
Comment (karrui) [inline src/app/modules/spcp/spcp.controller.ts:238]: add TODO to remove?

File: src/app/modules/spcp/spcp.service.ts [hunk 4]
```diff
@@ -96,14 +134,21 @@ export class SpcpService {
...
+    useCpCloud = false,
   ): Result<string, CreateRedirectUrlError> {
     const logMeta = {
       action: 'createRedirectUrl',
       authType,
       target,
       esrvcId,
     }
-    const authClient = this.getAuthClient(authType)
+    if (useCpCloud) {
```
Comment (karrui) [inline src/app/modules/spcp/spcp.service.ts:145]: add TODO to remove?

File: src/app/modules/submission/email-submission/email-submission.controller.ts [hunk 5]
```diff
@@ -126,11 +132,14 @@ export const handleEmailSubmission: RequestHandler<
   // Handle SingPass, CorpPass and MyInfo authentication and validation
   const { authType } = form
   if (authType === AuthType.SP || authType === AuthType.CP) {
+    const useCpCloud = spcpFeature.props?.cpCloudFormId === formId
```
Comment (karrui) [inline src/app/modules/submission/email-submission/email-submission.controller.ts:135]: add TODO to remove?

**Review by karrui — APPROVED** [review]
lgtm, maybe add more TODO markers, or we can just refer back to this PR

**Comment by mantariksh** [comment]
>lgtm, maybe add more TODO markers, or we can just refer back to this PR

I wrote the TODOs such that they will cascade, i.e. once you delete the code for the first TODO, you will know which variables are no longer defined so you know to delete them haha

**Pull request merged**